language/move-prover/bytecode/src/read_write_set_analysis.rs

```diff
@@ -0,0 +1,588 @@
+                    // These operations touch non-reference values; nothing to do
+                }
+                oper => panic!("unsupported oper {:?}", oper),
+            },
+            Load(_attr_id, lhs, constant) => {
+                if let Constant::Address(a) = constant {
+                    state.locals.bind_local(*lhs, AbsAddr::constant(a.clone()))
+                }
+            }
+            Assign(_attr_id, lhs, rhs, _assign_kind) => {
+                if let Some(rhs_data) = state.locals.get_local(*rhs).cloned() {
+                    state.locals.bind_local(*lhs, rhs_data)
+                } else {
+                    state.locals.remove_local(*lhs)
+                }
+            }
+            Ret(_attr_id, rets) => {
+                let ret_vals: Vec<Option<AbsAddr>> = rets
+                    .iter()
+                    .map(|ret| state.locals.get_local(*ret).cloned())
+                    .collect();
+                // clear all locals
+                state.locals.0.clear();
+                for (ret_index, ret_val_opt) in ret_vals.iter().enumerate() {
+                    if let Some(ret_val) = ret_val_opt {
+                        state.locals.bind_return(ret_index, ret_val.clone())
+                    }
+                }
+            }
+            Abort(..) => {
+                // clear all locals
+                state.locals.0.clear()
+            }
+            OnAbort(..) | SaveMem(..) | Prop(..) | SaveSpecVar(..) | Branch(..) | Jump(..)
+            | Label(..) | SpecBlock(..) | Nop(..) => (),
+        }
+    }
+}
+
+impl<'a> DataflowAnalysis for ReadWriteSetAnalysis<'a> {}
+impl<'a> CompositionalAnalysis<ReadWriteSetState> for ReadWriteSetAnalysis<'a> {
+    fn to_summary(&self, state: Self::State) -> ReadWriteSetState {
+        state
+    }
+}
+pub struct ReadWriteSetProcessor();
+impl ReadWriteSetProcessor {
+    pub fn new() -> Box<Self> {
+        Box::new(ReadWriteSetProcessor())
+    }
+}
+
+impl FunctionTargetProcessor for ReadWriteSetProcessor {
+    fn process(
+        &self,
+        targets: &mut FunctionTargetsHolder,
+        func_env: &FunctionEnv<'_>,
+        data: FunctionData,
+    ) -> FunctionData {
+        let fun_target = FunctionTarget::new(func_env, &data);
+        let mut initial_state = ReadWriteSetState::default();
+        // initialize_formals
+        for param_index in fun_target.get_parameters() {
+            initial_state
+                .locals
+                .bind_local(param_index, AbsAddr::formal(param_index))
+        }
+        let cache = SummaryCache::new(targets, func_env.module_env.env);
+        let analysis = ReadWriteSetAnalysis {
+            cache,
+            _func_env: func_env,
+        };
+        analysis.summarize(func_env, initial_state, data)
+    }
+
+    fn name(&self) -> String {
+        "read_write_set_analysis".to_string()
+    }
+}
+
+// ==== summary printing
```

Comment:
Suggest to make those section markers more systematic and nicer. Also maybe use sections more extensively to organize this and other new code. Please also add regular doc comments more exhaustively.

Reply:
Done